File: bindings/csharp/src/Session.cs
using System;
using Google.Protobuf;

namespace Angzarr.Router;

/// <summary>
/// One dispatch's host-side state object, reached from callbacks via
/// <c>host_ctx</c>. The rebuilt state is created lazily by the first stateful
/// callback (all callbacks in one dispatch share it). State is a mutable
/// <see cref="IMessage"/> (Google.Protobuf messages are mutable — no Builder);
/// appliers fold events into it and the handler reads it back.
/// </summary>
internal sealed class Session
{
    internal readonly Router Router;
    private IMessage? _state;

    internal Session(Router router) => Router = router;

    /// <summary>Lazily creates the host state from the factory on first callback,
    /// then reuses it across the dispatch.</summary>
    internal IMessage EnsureState(Func<IMessage> factory) => _state ??= factory();
}

File: bindings/typescript/src/session.ts
import { type Outcome } from "./statuses";

/**
 * One dispatch's host-side state object, reached from the callback trampoline
 * via its host_ctx id. The rebuilt state is created lazily by the first stateful
 * callback (all callbacks in one dispatch share it); state never crosses the
 * FFI. A session routes a callback to its router's registered invoker.
 */
export interface Session {
  /** Lazily creates the host state from the factory on first callback, then
   * reuses it across the dispatch. */
  ensureState<T>(factory: () => T): T;

  /** Routes one host callback to the registered invoker, marshaling inputs and
   * returning the response bytes + ABI status. Never throws — the firewall is
   * here, so a handler error becomes a coded negative status, never an unwind
   * across the FFI. */
  handleCallback(
    callbackId: number,
    typeUrl: string,
    payload: Uint8Array,
    aux: Uint8Array,
  ): Outcome;
}
